src/mame/taito/taito_f3_v.cpp

```diff
@@ -700,2153 +686,737 @@ u16 taito_f3_state::lineram_r(offs_t offset)
 
 void taito_f3_state::lineram_w(offs_t offset, u16 data, u16 mem_mask)
 {
-#if 0
-	/* DariusGX has an interesting bug at the start of Round D - the clearing of lineram
-	(0xa000->0x0xa7ff) overflows into priority RAM (0xb000) and creates garbage priority
-	values.  I'm not sure what the real machine would do with these values, and this
-	emulation certainly doesn't like it, so I've chosen to catch the bug here, and prevent
-	the trashing of priority ram.  If anyone has information on what the real machine does,
-	please let me know! */
-	/* Update: this doesn't seem to occur anymore, I'll leave this snippet in but commented out.
-	 *         fwiw PC=0x1768a0/0x1768a4 is where the game clears lineram in round D, which is a
-	 *         move.w Dn, (An,D7.w*2) , a kind of opcode that could've been bugged back then.
-	 */
-	if (m_game == DARIUSG)
-	{
-		if (m_f3_skip_this_frame)
-			return;
-		if (offset == 0xb000 / 2 && data == 0x003f)
-		{
-			m_f3_skip_this_frame = 1;
-			return;
-		}
-	}
-#endif
-
 	COMBINE_DATA(&m_line_ram[offset]);
 }
 
 void taito_f3_state::palette_24bit_w(offs_t offset, u32 data, u32 mem_mask)
 {
-	int r, g, b;
-
 	COMBINE_DATA(&m_paletteram32[offset]);
-
-	/* 12 bit palette games - there has to be a palette select bit somewhere */
-	if (m_game == SPCINVDX || m_game == RIDINGF || m_game == ARABIANM || m_game == RINGRAGE)
-	{
-		b = 15 * ((m_paletteram32[offset] >> 4) & 0xf);
-		g = 15 * ((m_paletteram32[offset] >> 8) & 0xf);
-		r = 15 * ((m_paletteram32[offset] >> 12) & 0xf);
+	const u32 color = m_paletteram32[offset];
+	rgb_t rgb;
+
+	/* TODO: 12 bit palette games - seems to be selected on a line basis by 6400? */
+	if (m_game == SPCINVDX || m_game == RIDINGF || m_game == ARABIANM || m_game == RINGRAGE) {
+		// .... .... .... .... RRRR GGGG BBBB ....
+		// .... .... RRRR rrrr GGGG gggg BBBB bbbb
+		rgb = rgb_t(BIT(color, 12, 4) * 16, BIT(color, 8, 4) * 16, BIT(color, 4, 4) * 16);
+	} else {
+		rgb = rgb_t(color).set_a(255);
 	}
 
-	/* This is weird - why are only the sprites and VRAM palettes 21 bit? */
-	else if (m_game == CLEOPATR)
+	m_palette->set_pen_color(offset, rgb);
+}
+
+/******************************************************************************/
+
+// line: [latched] line info from previous call, will modify in-place
+// y should be called 0->255 for non-flipscreen, 255->0 for flipscreen
+void taito_f3_state::read_line_ram(f3_line_inf &line, int y)
+{
+	const auto latched_addr = [=] (u8 section, u8 subsection) -> offs_t
 	{
```

Comment:
Please don’t use auto capture in lambdas – it’s better to be explicit abut what you want to capture.

Reply:
done